src/main/kotlin/uk/gov/justice/digital/hmpps/visitscheduler/service/repoting/VisitsReportingService.kt

```diff
@@ -0,0 +1,160 @@
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.PrisonDto
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.reporting.SessionVisitCountsDto
+import uk.gov.justice.digital.hmpps.visitscheduler.dto.sessions.SessionScheduleDto
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitRestriction
+import uk.gov.justice.digital.hmpps.visitscheduler.model.VisitStatus
+import uk.gov.justice.digital.hmpps.visitscheduler.model.entity.projections.VisitRestrictionStats
+import uk.gov.justice.digital.hmpps.visitscheduler.repository.VisitRepository
+import uk.gov.justice.digital.hmpps.visitscheduler.service.PrisonConfigService
+import uk.gov.justice.digital.hmpps.visitscheduler.service.SessionService
+import java.time.LocalDate
+
+@Service
+class VisitsReportingService(
+  private val prisonConfigService: PrisonConfigService,
+  private val sessionService: SessionService,
+  private val visitRepository: VisitRepository,
+) {
+  fun getVisitCountsBySession(reportDate: LocalDate): List<SessionVisitCountsDto> {
+    return getSessionsReport(reportDate)
+  }
+
+  private fun getSessionsReport(reportDate: LocalDate): List<SessionVisitCountsDto> {
+    val prisons = getAllActivePrisons()
+    return getSessionVisitCounts(reportDate, getVisitCountsForPrisons(prisons, reportDate))
+  }
+
+  private fun getVisitCountsForPrisons(prisons: List<PrisonDto>, reportDate: LocalDate): Map<PrisonDto, Map<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>>> {
+    val prisonDetails = mutableMapOf<PrisonDto, Map<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>>>()
+    prisons.forEach { prison ->
+      prisonDetails[prison] = getVisitCountsForPrison(prison, reportDate)
+    }
+
+    return prisonDetails
+  }
+
+  private fun getVisitCountsForPrison(prison: PrisonDto, reportDate: LocalDate): Map<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>> {
+    var sessionDetails = mapOf<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>>()
+    val isExcludedDate = isExcludedDate(prison, reportDate)
+
+    if (!isExcludedDate) {
+      val sessions = getSessionsByDateForPrison(prison, reportDate)
+      sessionDetails = getVisitCountsForSessions(sessions, reportDate)
+    }
+
+    return sessionDetails
+  }
+
+  private fun isExcludedDate(prison: PrisonDto, reportDate: LocalDate): Boolean {
+    return prison.excludeDates.contains(reportDate)
+  }
+
+  private fun getVisitCountsForSessions(sessions: List<SessionScheduleDto>, reportDate: LocalDate): Map<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>> {
+    val sessionDetails = mutableMapOf<SessionScheduleDto, Map<Pair<VisitStatus, VisitRestriction>, Int>>()
+    sessions.forEach { session ->
+      sessionDetails[session] = getVisitCountsForSession(session, reportDate)
+    }
+
+    return sessionDetails
+  }
+
+  private fun getAllActivePrisons(): List<PrisonDto> {
+    return prisonConfigService.getPrisons().filter { it.active }
```

Comment:
Should use prisonConfigService.getSupportedPrisons()

Reply:
getSupportedPrisons() returns a list of Strings while I need the PrisonDtos back. Should I add a new method on the PrisonConfigService as getActivePrisons(): List<PrisonDto>?

Reply:
I think that getSupportedPrisons prisons is :
  @Query(
    "select p.code from Prison p " +
      "where p.active = true order by p.code",
  )
  fun getSupportedPrisons(): List<String>

could call it differently